[WIP] Turned on -Wconversion for diff.c

wreed4 opened the pull request:
Documenting that I have chosen to work on diff.c  for issue #567 

There seems to be quite a lot of overlap with mouse.c

wreed4 committed:
Turned on -Wconversion for diff.c

wreed4 committed:
intermittent commit

made a lot of things long.  Tried not to cast too much.  Failed somewhat
in some places.

wreed4 commented:
I seem to basically have two options here: make lots of things `long`, or make `linenr_T` an `int` (and thus a lot of other stuff `int` also).  I think this second option is better, mostly because a line number makes more sense as an `int` than a `long`.  Vim has trouble enough opening a file with upwards of 30-50K lines.  I shudder to think of trying to open (or find..or see stored on a disk..) a file with more than 2billion lines (still under the size of a 32-bit int).

splinterofchaos commented:
> I seem to basically have two options here: make lots of things long, or make linenr_T an int

I haven't examined the situation, myself, but what about making lots of things `linenr_T`? Also, do we ever use a `linenr_T` for a negative value? Could it be `unsigned`? The idea of changing it to `int` makes sense, but there might be historical context to making it a `long`. Maybe even `size_t` would be more appropriate.

wreed4 commented:
I did make as many things as appropriate `linenr_T`.  I tried to avoid "raw" types (long, int, etc) as much as I could, applying previously defined typedefs where ever appropriate.  And I did toy around with the idea of making them unsigned.  However, there is at least one place I found the use of a negative line number in movements (think `-4d`).  This seems to be a valid scenario and enough reason (in my mind) to just leave it as a plain `int`.  But if you (or anyone else) disagrees, please let me know :)

wreed4 committed:
No warnings!!!

changed `linenr_T` to be `int`. This makes more conseptual sense, as I
sincerely doubt any file will be longer than 2^31-1 lines long. Also,
the macro `MAXLNUM` was already defined to be a 31-bit value.

wreed4 committed:
Fixed some errors that didn't arise on my machine

wreed4 committed:
Minor formatting fix

splinterofchaos commented on src/nvim/buffer_defs.h:
```diff
@@ -965,8 +965,8 @@ struct window_S {
-  int w_topfill;                    /* number of filler lines above w_topline */
-  int w_old_topfill;                /* w_topfill at last redraw */
+  linenr_T w_topfill;                    /* number of filler lines above w_topline */
+  linenr_T w_old_topfill;                /* w_topfill at last redraw */
```
Please try to keep documentation aligned, and convert to doxygen style (`///< w_topfil at last redraw`).

wreed4 replied:
Thanks, good catch :).

splinterofchaos commented on src/nvim/diff.c:
```diff
@@ -952,7 +952,7 @@ void ex_diffpatch(exarg_T *eap)
     if (curbuf->b_fname != NULL) {
       newname = vim_strnsave(curbuf->b_fname,
-                             (int)(STRLEN(curbuf->b_fname) + 4));
+                             (size_t)(STRLEN(curbuf->b_fname) + 4));
```
Is this cast needed? `strlen()` returns a `size_t` and I don't think `+ 4` will make it signed.

splinterofchaos commented on src/nvim/diff.c:
```diff
@@ -1425,11 +1425,11 @@ void diff_clear(tabpage_T *tp)
 ///
 /// @return diff status.
-int diff_check(win_T *wp, linenr_T lnum)
+linenr_T diff_check(win_T *wp, linenr_T lnum)
```
I think `int` would be more appropriate here since the result may not be related to a line number.

wreed4 commented:
I'm not sure why these builds are failing.  I go through the log (particularly on the clang Travis build) and it says success...but it is marked as failure because it returned 1 for some reason.  The only errors I see are errors that pip is out of date.  Any thoughts?

elmart commented on src/nvim/api/buffer.c:
```diff
@@ -193,12 +193,14 @@ void buffer_set_line_slice(Buffer buffer,
-  ssize_t extra = 0;  // lines added to text, can be negative
-  char **lines = (new_len != 0) ? xcalloc(new_len, sizeof(char *)) : NULL;
-
-  for (size_t i = 0; i < new_len; i++) {
```
Sorry, but I don't agree to all this changing size_t to other types.
All these had been previously refactored to `size_t`, as it is the natural type for sizes, lenghts, counts, indexes, etc.
Changing all of them just because at some point they interact with a `linenr_T` is not what desired.
Better to (safely) cast the involved `linenr_T` when needed.
I write this here cause it's the first occurrence, but what I said applies to many other changes after this.

splinterofchaos commented:
Don't worry about the failures. quickbuild have been failing often, and travis does fail occasionally. For travis, if only one test fails (unless the log has a message from clang ASAN), it's generally not a problem.

Pyrohh commented:
I'll close this for now, but please re-open if you plan on working on it more.
